File: back/EdtechBack/src/main/java/pi/tn/esprit/services/CertificateService.java
package pi.tn.esprit.services;

import pi.tn.esprit.models.Certificate;
import pi.tn.esprit.models.Classe;
import pi.tn.esprit.models.Comment;

import java.util.List;

public interface CertificateService {
    List<Certificate> retrieveAllCertificates();
    Certificate retrieveCertificate(int certificateId);
    Certificate addCertificate(Certificate certificate);
    void removeCertificate(int certificateId);
    Certificate modifyCertificate(Certificate certificate);

    void assignMajorToCertificate(int certificateId, int majorId);

    void deassignMajorFromCertificate(int certificateId);
     List<Certificate> getCertificatesByCategory(String category);
     Certificate addRating(int certificateId, int rating);
     Comment addComment(int certificateId, String username, String text);
    public List<Comment> getComments(int certificateId);
}

File: front/EdTech/src/app/service/certificate.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Certificate } from '../model/certificate';
import { Comment } from '../model/comment';

@Injectable({
  providedIn: 'root',
})
export class CertificateService {
  private baseUrl = 'http://localhost:7777/api/certificates';
  private baseUrl2 = 'http://localhost:7777/api/students';

  constructor(private http: HttpClient) {}

  retrieveAllCertificates(): Observable<Certificate[]> {
    return this.http.get<Certificate[]>(`${this.baseUrl}/retrieve-all`);
  }
  getCertificatesByCategory(category: string): Observable<Certificate[]> {
    return this.http.get<Certificate[]>(
      `${this.baseUrl}/retrieve-by-category/${category}`
    );
  }

  getCertificateById(id: number): Observable<Certificate> {
    return this.http.get<Certificate>(`${this.baseUrl}/retrieve/${id}`);
  }

  addCertificate(certificate: Certificate): Observable<Certificate> {
    return this.http.post<Certificate>(`${this.baseUrl}/add`, certificate);
  }

  removeCertificate(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/remove/${id}`);
  }

  modifyCertificate(certificate: Certificate): Observable<Certificate> {
    return this.http.put<Certificate>(`${this.baseUrl}/modify`, certificate);
  }
  getComments(certificateId: number): Observable<Comment[]> {
    return this.http.get<Comment[]>(
      `${this.baseUrl}/${certificateId}/comments`
    );
  }

  addComment(certificateId: number, comment: Comment): Observable<Comment> {
    return this.http.post<Comment>(
      `${this.baseUrl}/${certificateId}/comments`,
      comment
    );
  }

  addRating(certificateId: number, rating: number): Observable<Certificate> {
    console.log(rating);
    return this.http.post<Certificate>(
      `${this.baseUrl}/${certificateId}/ratings`,
      rating
    );
  }
  getCertificates(studentId: number): Observable<Certificate[]> {
    return this.http.get<Certificate[]>(
      `${this.baseUrl2}/getcertif/${studentId}`
    );
  }
}
